### ISSUE TITLE
Config for Hashes of objects seems to not be working.

### ISSUE THREAD

**Issue by juzham** [mdub/config_mapper#5]
Hi @mdub,

Hope you are going well! Cheers for the work you have put into maintaining this gem.
Was inspired to use it after your chat at REA Ruby guild.

Only thing is I am having trouble running the example in the README for mapping nested config.

```
require 'config_mapper'
require 'config_mapper/config_struct'

class Position < ConfigMapper::ConfigStruct

  attr_reader :x
  attr_reader :y

  def x=(arg); @x = Integer(arg); end
  def y=(arg); @y = Integer(arg); end

end

positions = Hash.new { |h,k| h[k] = Position.new }

config_data = {
  "fred" => { "x" => 2, "y" => 4 },
  "mary" => { "x" => 3, "y" => 5 }
}

ConfigMapper.configure_with(config_data, positions)
puts positions["fred"].x            #undefined method `x' for {"x"=>2, "y"=>4}:Hash (NoMethodError)
puts positions["mary"].y
```

If I use version 1.2 I can get it to work, so seems like this changed from version 1.2 to 1.3

These are the two release commits
```
git diff 007d777a1e44926337db41179770ed96c68e47f4 18fb8aa381ec54909404aa98aee258e51e4d76a1
```

Seems like the noteable change was from using `dict_mapper` to `collection_mapper` for hashs.

Unsure if nested config was purposely removed or if this is a bug.

**Comment by mdub** [comment]
Hi @juzham. Hmm, that should work ... I'll have to look into it.

Looks like you're mashing up two different parts of the README. If you're using `ConfigStruct`, you can declare config attributes using it's DSL, e.g.

```ruby
class Config < ConfigMapper::ConfigStruct

  component_dict :positions do
    attribute :x, Integer
    attribute :y, Integer
  end

end
```

And populate it like this:

```ruby
config_data = {
  "positions" => {
    "fred" => { "x" => 2, "y" => 4 },
    "mary" => { "x" => 3, "y" => 5 }
  } 
}

config = Config.from_data(config_data)
config.positions["fred"].x #=> 2
```

**Comment by juzham** [comment]
Thanks @mdub the above works and will solve the problem I am trying to solve.
(and yes was mashing concepts before)
If I am reading the README correctly, I think this should work as well though?
```ruby
require 'config_mapper'

class Position

  attr_reader :x
  attr_reader :y

  def x=(arg); @x = Integer(arg); end
  def y=(arg); @y = Integer(arg); end

end

positions = Hash.new { |h,k| h[k] = Position.new }

config_data = {
  "fred" => { "x" => 2, "y" => 4 },
  "mary" => { "x" => 3, "y" => 5 }
}

ConfigMapper.configure_with(config_data, positions)
puts positions["fred"].x   # undefined method `x' for {"x"=>2, "y"=>4}:Hash (NoMethodError)
```

**Comment by mdub** [comment]
Yeah, that _should_ work. I’ll have to break out the debugger.

**Comment by mdub** [comment]
Actually, you're right, @juzham: the behaviour _did_ change in version 1.3. The README is now inaccurate.

Previously, the logic for setting an attribute was something like:

```ruby
def configure_attribute(key, value)
  if value.is_a?(Hash) && !get(key).nil?
    ConfigMapper.configure_with(value, get(key)) 
  else
    set(key, value)
  end
end
```

But, the type testing was causing problems, so I moved to a more duck-typed approach, more like:

```ruby
def configure_attribute(key, value)
  if can_set?(key)
    set(key, value)
  else
    ConfigMapper.configure_with(value, get(key)) 
  end
end
```

In other words, if an attribute "`a`" is directly set-able (because the target has an `#a=` method, _or_ a `#[]=` method), we simply set it. Otherwise, we access it's value (by calling `.a` or `["a"]`) and try to map data onto _that_.

I'm thinking I'll call this a documentation bug. I need some documentation around `component_dict`, anyway.
